File: bindings/swift/Sources/MaplibreNativeFFI/Camera.swift
internal import CMaplibreNativeC

public struct CameraOptions: Equatable, Sendable {
  public var center: LatLng?
  public var zoom: Double?
  public var bearing: Double?
  public var pitch: Double?
  public var centerAltitude: Double?
  public var padding: EdgeInsets?
  /// Input-only screen point the camera pivots around. Camera updates honor it;
  /// MapLibre leaves it `nil` on camera reads.
  public var anchor: ScreenPoint?
  public var roll: Double?
  public var fieldOfView: Double?

  public init(
    center: LatLng? = nil,
    zoom: Double? = nil,
    bearing: Double? = nil,
    pitch: Double? = nil,
    centerAltitude: Double? = nil,
    padding: EdgeInsets? = nil,
    anchor: ScreenPoint? = nil,
    roll: Double? = nil,
    fieldOfView: Double? = nil
  ) {
    self.center = center
    self.zoom = zoom
    self.bearing = bearing
    self.pitch = pitch
    self.centerAltitude = centerAltitude
    self.padding = padding
    self.anchor = anchor
    self.roll = roll
    self.fieldOfView = fieldOfView
  }

  init(native: NativeCameraOptionsInput) {
    center = native.center.map(LatLng.init(native:))
    zoom = native.zoom
    bearing = native.bearing
    pitch = native.pitch
    centerAltitude = native.centerAltitude
    padding = native.padding.map(EdgeInsets.init(native:))
    anchor = native.anchor.map { ScreenPoint(x: $0.x, y: $0.y) }
    roll = native.roll
    fieldOfView = native.fieldOfView
  }

  var nativeInput: NativeCameraOptionsInput {
    NativeCameraOptionsInput(
      center: center?.nativeInput,
      zoom: zoom,
      bearing: bearing,
      pitch: pitch,
      centerAltitude: centerAltitude,
      padding: padding?.nativeInput,
      anchor: anchor?.nativeInput,
      roll: roll,
      fieldOfView: fieldOfView
    )
  }
}

public struct UnitBezier: Equatable, Sendable {
  public var x1: Double
  public var y1: Double
  public var x2: Double
  public var y2: Double

  public init(x1: Double, y1: Double, x2: Double, y2: Double) {
    self.x1 = x1
    self.y1 = y1
    self.x2 = x2
    self.y2 = y2
  }

  var nativeInput: NativeUnitBezier {
    NativeUnitBezier(x1: x1, y1: y1, x2: x2, y2: y2)
  }
}

public struct AnimationOptions: Equatable, Sendable {
  public var durationMilliseconds: Double?
  public var velocity: Double?
  public var minimumZoom: Double?
  public var easing: UnitBezier?
  /// Caller-chosen identity for the transition these options start. When
  /// present, the transition emits one `.mapCameraTransitionFinished` runtime
  /// event carrying this value; when absent, it emits none.
  ///
  /// The event fires exactly once however the transition ends, including when a
  /// later camera command supersedes it, and reports no outcome. A rejected
  /// command starts no transition and emits no event.
  public var transitionId: UInt64?

  public init(
    durationMilliseconds: Double? = nil,
    velocity: Double? = nil,
    minimumZoom: Double? = nil,
    easing: UnitBezier? = nil,
    transitionId: UInt64? = nil
  ) {
    self.durationMilliseconds = durationMilliseconds
    self.velocity = velocity
    self.minimumZoom = minimumZoom
    self.easing = easing
    self.transitionId = transitionId
  }

  var nativeInput: NativeAnimationOptionsInput {
    NativeAnimationOptionsInput(
      durationMilliseconds: durationMilliseconds,
      velocity: velocity,
      minimumZoom: minimumZoom,
      easing: easing?.nativeInput,
      transitionId: transitionId
    )
  }
}

public enum CameraUpdateMode: UInt32, Sendable, Hashable {
  case jump = 0
  case ease = 1
  case fly = 2
}

public enum GesturePhase: UInt32, Sendable, Hashable {
  case none = 0
  case begin = 1
  case update = 2
  case end = 3
  case cancel = 4
}

/// One camera command copied atomically by the runtime worker.
public struct CameraUpdate: Equatable, Sendable {
  public var mode: CameraUpdateMode
  public var camera: CameraOptions
  public var animation: AnimationOptions
  public var gesturePhase: GesturePhase
  public var gestureId: UInt64
  public var animationId: UInt64

  public init(
    mode: CameraUpdateMode = .jump,
    camera: CameraOptions,
    animation: AnimationOptions = AnimationOptions(),
    gesturePhase: GesturePhase = .none,
    gestureId: UInt64 = 0,
    animationId: UInt64 = 0
  ) {
    self.mode = mode
    self.camera = camera
    self.animation = animation
    self.gesturePhase = gesturePhase
    self.gestureId = gestureId
    self.animationId = animationId
  }

  func withNativeUpdate<Result>(
    _ body: (UnsafePointer<mln_camera_update>) throws -> Result
  ) throws -> Result {
    try camera.nativeInput.withNativeOptions { camera in
      try animation.nativeInput.withOptionalNativeOptions { animation in
        var update = mln_camera_update_default()
        update.mode = mode.rawValue
        update.camera = camera.pointee
        if let animation {
          update.animation = animation.pointee
        }
        update.gesture_phase = gesturePhase.rawValue
        update.gesture_id = gestureId
        update.animation_id = animationId
        return try withUnsafePointer(to: &update, body)
      }
    }
  }
}

File: bindings/python/python/maplibre_native_ffi/camera.py
"""Camera descriptors and map camera operations."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

from .geo import LatLng, LatLngBounds


@dataclass(frozen=True, slots=True)
class ScreenPoint:
    """Screen-space point in logical map pixels."""

    x: float
    y: float


@dataclass(frozen=True, slots=True)
class EdgeInsets:
    """Screen-space inset in logical map pixels."""

    top: float = 0.0
    left: float = 0.0
    bottom: float = 0.0
    right: float = 0.0


@dataclass(frozen=True, slots=True)
class UnitBezier:
    """Cubic Bezier control points for animation easing."""

    p1x: float
    p1y: float
    p2x: float
    p2y: float


@dataclass(frozen=True, slots=True)
class Vec3:
    """Three-component vector used by free camera options."""

    x: float
    y: float
    z: float


@dataclass(frozen=True, slots=True)
class Quaternion:
    """Quaternion stored as x, y, z, w components."""

    x: float
    y: float
    z: float
    w: float


@dataclass(frozen=True, slots=True)
class CameraOptions:
    """Camera fields used for snapshots and camera commands."""

    center: LatLng | None = None
    zoom: float | None = None
    bearing: float | None = None
    pitch: float | None = None
    center_altitude: float | None = None
    padding: EdgeInsets | None = None

    anchor: ScreenPoint | None = None
    """Screen-space anchor for jump, ease, and fly commands. Input-only: a
    camera read always reports None."""

    roll: float | None = None
    field_of_view: float | None = None

    @classmethod
    def _from_native(cls, raw: dict[str, object]) -> CameraOptions:
        center = raw["center"]
        padding = raw["padding"]
        anchor = raw["anchor"]
        return cls(
            center=LatLng(**center) if isinstance(center, dict) else None,
            zoom=raw["zoom"],
            bearing=raw["bearing"],
            pitch=raw["pitch"],
            center_altitude=raw["center_altitude"],
            padding=EdgeInsets(**padding) if isinstance(padding, dict) else None,
            anchor=ScreenPoint(**anchor) if isinstance(anchor, dict) else None,
            roll=raw["roll"],
            field_of_view=raw["field_of_view"],
        )


@dataclass(frozen=True, slots=True)
class AnimationOptions:
    """Optional animation controls for camera transitions.

    ``transition_id`` is a caller-chosen identity passed through uninterpreted.
    When set, the transition reports its end once through a
    ``MAP_CAMERA_TRANSITION_FINISHED`` runtime event carrying that value.
    """

    duration_ms: float | None = None
    velocity: float | None = None
    min_zoom: float | None = None
    easing: UnitBezier | None = None
    transition_id: int | None = None


class CameraDeltaKind(IntEnum):
    """Relative camera operation kind."""

    MOVE = 0
    SCALE = 1
    BEARING = 2
    PITCH = 3


@dataclass(frozen=True, slots=True)
class CameraDelta:
    """One relative camera operation."""

    kind: CameraDeltaKind = CameraDeltaKind.MOVE
    offset: ScreenPoint = ScreenPoint(0.0, 0.0)
    amount: float = 0.0
    anchor: ScreenPoint | None = None
    animation: AnimationOptions = AnimationOptions()


@dataclass(frozen=True, slots=True)
class CameraFitOptions:
    """Optional fitting controls for camera-for-viewport queries."""

    padding: EdgeInsets | None = None
    bearing: float | None = None
    pitch: float | None = None


@dataclass(frozen=True, slots=True)
class Bounded:
    """Keeps the camera center inside the given bounds."""

    bounds: LatLngBounds


@dataclass(frozen=True, slots=True)
class Unbounded:
    """Leaves the camera center unconstrained.

    The map pans freely across the antimeridian. This differs from world bounds
    of -90/-180 to 90/180, which clamp longitude to that range.
    """


type BoundsConstraint = Bounded | Unbounded


@dataclass(frozen=True, slots=True)
class BoundOptions:
    """Optional map camera constraint fields."""

    bounds: BoundsConstraint | None = None
    min_zoom: float | None = None
    max_zoom: float | None = None
    min_pitch: float | None = None
    max_pitch: float | None = None

    @classmethod
    def _from_native(cls, raw: dict[str, object]) -> BoundOptions:
        raw_bounds = raw["bounds"]
        bounds: BoundsConstraint | None
        if isinstance(raw_bounds, dict):
            bounds = Bounded(
                LatLngBounds(
                    southwest=LatLng(**raw_bounds["southwest"]),
                    northeast=LatLng(**raw_bounds["northeast"]),
                )
            )
        elif raw["unbounded"]:
            bounds = Unbounded()
        else:
            bounds = None
        return cls(
            bounds=bounds,
            min_zoom=raw["min_zoom"],
            max_zoom=raw["max_zoom"],
            min_pitch=raw["min_pitch"],
            max_pitch=raw["max_pitch"],
        )


@dataclass(frozen=True, slots=True)
class FreeCameraOptions:
    """Free camera position and orientation in MapLibre camera space."""

    position: Vec3 | None = None
    orientation: Quaternion | None = None

    @classmethod
    def _from_native(cls, raw: dict[str, object]) -> FreeCameraOptions:
        position = raw["position"]
        orientation = raw["orientation"]
        return cls(
            position=Vec3(**position) if isinstance(position, dict) else None,
            orientation=Quaternion(**orientation)
            if isinstance(orientation, dict)
            else None,
        )


@dataclass(frozen=True, slots=True)
class ProjectionMode:
    """Axonometric rendering options for the live map render transform."""

    axonometric: bool | None = None
    x_skew: float | None = None
    y_skew: float | None = None

    @classmethod
    def _from_native(cls, raw: dict[str, object]) -> ProjectionMode:
        return cls(
            axonometric=raw["axonometric"],
            x_skew=raw["x_skew"],
            y_skew=raw["y_skew"],
        )


__all__ = [
    "AnimationOptions",
    "BoundOptions",
    "Bounded",
    "BoundsConstraint",
    "CameraFitOptions",
    "CameraOptions",
    "EdgeInsets",
    "FreeCameraOptions",
    "ProjectionMode",
    "Quaternion",
    "ScreenPoint",
    "Unbounded",
    "UnitBezier",
    "Vec3",
]
